dhdt/auxilary/handler_mgrs.py

```diff
@@ -0,0 +1,142 @@
+import pandas as pd
+supported_drivers['KML'] = 'rw'
+
+MGRS_TILING_URL = (
+    "https://sentinels.copernicus.eu/documents/247904/1955685/"
+    "S2A_OPER_GIP_TILPAR_MPC__20151209T095117_V20150622T000000"
+    "_21000101T000000_B00.kml"
+)
+
+MGRS_TILING_DIR_DEFAULT = os.path.join('.', 'data', 'MGRS')
+
+
+def download_mgrs_tiling(mgrs_dir=None, output='sentinel2_tiles_world.geojson', keep_kml=False):
+    """
+    Retrieve mgrs tiling polygons. 
+
+    Parameters
+    ----------
+    mgrs_dir : str, optional
+        location where the MGRS tiling files will be saved. If None, files will be written to './data/MGRS'. By default None
+    output: str, optional
+        Output file name, by default 'sentinel2_tiles_world.geojson'
+    keep_kml : bool, optional
+        If keep the kml format. If False, will convert the kml data to geojson. By default False
+
+    Returns
+    -------
+    mgrs_path : str
+        path to the extracted MGRS tiling file
+
+    Notes
+    -----
+    KML file with the MGRS tiling scheme used by Sentinel-2 is provided by
+    Copernicus [1]
+
+    [1] https://sentinels.copernicus.eu/web/sentinel/missions/sentinel-2/data-products
+    """
+
+    mgrs_dir = MGRS_TILING_DIR_DEFAULT if mgrs_dir is None else mgrs_dir
+
+    # download MGRS tiling file
+    f_kml = get_file_from_www(MGRS_TILING_URL, mgrs_dir)
+
+    if not keep_kml:
+        f_geojson = os.path.join(mgrs_dir, output)
+        gdf = _kml_to_gdf(os.path.join(mgrs_dir, f_kml))
+        gdf.to_file(f_geojson)
+        return f_geojson
+    else:
+        return f_kml
+
+
+def _kml_to_gdf(filename):
+    """
+    Convert MGRS kml file to a GeoPandas DataFrame
+
+    Parameters
+    ----------
+    filename : str
+        kml file name
+
+    Returns
+    -------
+    str
+        gejson file name
+    """
+    # Load kml file
+    gdf = gpd.read_file(filename, driver="KML")
+
+    # Separate the entries crossing the antimeridian
+    mask_cross = gdf["geometry"].apply(_check_cross)
+    gdf_cross = gdf.loc[mask_cross].copy()
+    gdf_normal = gdf.loc[~mask_cross].copy()
+
+    # Read in normal polygons
+    # Their geometry will contain two elements, the first one always footprint polygon
+    # The second one is likely to be the centroid, which will be disgarded
+    gdf_normal.update(gdf_normal["geometry"].apply(lambda x: x.geoms[0]))
+    gdf_normal = gdf_normal.set_crs(gdf.crs)
+
+    # Read polygons crossing the crossing the antimeridian
+    # Their geometry will contain the three elements
+    # The second or third element will be another polygon, which will be read as a separate row
+    unpacked = gdf_cross.apply(_unpack_mgrs_geom, axis=1, result_type='expand')
+    gs_cross_list = pd.concat([unpacked[idx]
+                              for idx in unpacked.columns]).to_list()
+    gdf_cross = gpd.GeoDataFrame(gs_cross_list)
+    gdf_cross = gdf_cross.set_crs(gdf.crs)
+
+    gdf = pd.concat([gdf_cross, gdf_normal])
```

Comment:
I was not aware of the following, but the `explode` method of a `GeoDataFrame` unpack multi-part geometries, see [docs](https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoDataFrame.explode.html). So, for example:

```python
gdf_unpacked_geoms = gdf.explode(index_parts=False)
print(gdf_unpacked_geoms.geometry)
```
```
0        POLYGON Z ((180.00000 -73.05974 0.00000, 176.8...
0                    POINT Z (178.43231 -72.53611 0.00000)
0        POLYGON Z ((-180.00000 -72.07333 0.00000, -179...
1        POLYGON Z ((180.00000 -83.80855 0.00000, 174.7...
1                    POINT Z (177.35644 -83.30263 0.00000)
                               ...                        
56684    POLYGON Z ((-180.00000 81.03542 0.00000, -176....
56684                POINT Z (-178.34842 80.53223 0.00000)
56685    POLYGON Z ((180.00000 80.94781 0.00000, 176.99...
56685    POLYGON Z ((-180.00000 81.93139 0.00000, -176....
56685                POINT Z (-178.00336 81.42584 0.00000)
Name: geometry, Length: 113670, dtype: geometry
```
We could simply do `explode` on `gdf`, then mask out the rows whose geometry is not a polygon, and we are done. What do you think?